src/state.rs

```diff
@@ -0,0 +1,219 @@
+// Copyright 2022 MaidSafe.net limited.
+//
+// This SAFE Network Software is licensed to you under The General Public License (GPL), version 3.
+// Unless required by applicable law or agreed to in writing, the SAFE Network Software distributed
+// under the GPL Licence is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied. Please review the Licences for the specific language governing
+// permissions and limitations relating to use of the SAFE Network Software.
+
+use bls::{PublicKey, PublicKeySet, SecretKey, SecretKeyShare};
+use std::collections::{BTreeMap, BTreeSet};
+use std::sync::Arc;
+
+use crate::error::{Error, Result};
+use crate::knowledge::{Knowledge, KnowledgeFault};
+use crate::sdkg::{AckOutcome, Part, PartOutcome, SyncKeyGen};
+use crate::vote::{DkgSignedVote, DkgVote, IdAck, IdPart, NodeId};
+
+/// State of the Dkg session, contains the sync keygen and currently known Parts and Acks
+/// Can handle votes coming from other participants
+pub struct DkgState<R: bls::rand::RngCore> {
+    id: NodeId,
+    secret_key: SecretKey,
+    pub_keys: BTreeMap<NodeId, PublicKey>,
+    keygen: SyncKeyGen<NodeId>,
+    our_part: Part,
+    all_votes: BTreeSet<DkgSignedVote>,
+    rng: R,
+}
+
+pub enum VoteResponse {
+    WaitingForMoreVotes,
+    BroadcastVote(Box<DkgSignedVote>),
+    RequestAntiEntropy,
+    AntiEntropy(BTreeSet<DkgSignedVote>),
+    DkgComplete(PublicKeySet, SecretKeyShare),
+}
+
+enum DkgCurrentState {
+    IncompatibleVotes,
+    NeedAntiEntropy,
+    Termination(BTreeMap<IdPart, BTreeSet<IdAck>>),
+    WaitingForTotalAgreement,
+    GotAllAcks(BTreeMap<IdPart, BTreeSet<IdAck>>),
+    WaitingForMoreAcks,
+    GotAllParts(BTreeSet<IdPart>),
+    WaitingForMoreParts,
+}
+
+impl<R: bls::rand::RngCore + Clone> DkgState<R> {
+    pub fn new(
+        our_id: NodeId,
+        secret_key: SecretKey,
+        pub_keys: BTreeMap<NodeId, PublicKey>,
+        threshold: usize,
+        rng: &mut R,
+    ) -> Result<Self> {
+        let (sync_key_gen, opt_part) = SyncKeyGen::new(
+            our_id,
+            secret_key.clone(),
+            Arc::new(pub_keys.clone()),
```

Comment:
We shouldn't need the `Arc` on these pub keys. If it's run in a multi-threaded context, then the caller can wrap the entire `DkgState` in an Arc themselves.